Simperium/src/main/java/name/fraser/neil/plaintext/diff_match_patch.java

```diff
@@ -1472,6 +1475,92 @@ public String diff_toDelta(LinkedList<Diff> diffs) {
 
+  private int digit16(char c) throws IllegalArgumentException {
+    switch (c) {
+      case '0': return 0;
+      case '1': return 1;
+      case '2': return 2;
+      case '3': return 3;
+      case '4': return 4;
+      case '5': return 5;
+      case '6': return 6;
+      case '7': return 7;
+      case '8': return 8;
+      case '9': return 9;
+      case 'A': case 'a': return 10;
+      case 'B': case 'b': return 11;
+      case 'C': case 'c': return 12;
+      case 'D': case 'd': return 13;
+      case 'E': case 'e': return 14;
+      case 'F': case 'f': return 15;
+      default: throw new IllegalArgumentException();
+    }
+  }
+
+  private String decodeURI(String text) throws IllegalArgumentException {
+    int i = 0;
+    StringBuilder decoded = new StringBuilder(text.length());
+    while (i < text.length()) {
+      if (text.charAt(i) != '%') {
+        decoded.append(text.charAt(i++));
+        continue;
+      }
+      // start a percent-sequence
+      int byte1 = (digit16(text.charAt(i + 1)) << 4) + digit16(text.charAt(i + 2));
+      if ((byte1 & 0x80) == 0) {
+        decoded.append(Character.toChars(byte1));
+        i += 3;
+        continue;
+      }
+      if (text.charAt(i + 3) != '%') {
+        throw new IllegalArgumentException();
+      }
+      int byte2 = (digit16(text.charAt(i + 4)) << 4) + digit16(text.charAt(i + 5));
+      if ((byte2 & 0xC0) != 0x80) {
+        throw new IllegalArgumentException();
+      }
+      byte2 = byte2 & 0x3F;
+      if ((byte1 & 0xE0) == 0xC0) {
+        decoded.append(Character.toChars(((byte1 & 0x1F) << 6) | byte2));
+        i += 6;
+        continue;
+      }
+      if (text.charAt(i + 6) != '%') {
+        throw new IllegalArgumentException();
+      }
+      int byte3 = (digit16(text.charAt(i + 7)) << 4) + digit16(text.charAt(i + 8));
+      if ((byte3 & 0xC0) != 0x80) {
+        throw new IllegalArgumentException();
+      }
+      byte3 = byte3 & 0x3F;
+      if ((byte1 & 0xF0) == 0xE0) {
+        // unpaired surrogate are fine here
+        decoded.append(Character.toChars(((byte1 & 0x0F) << 12) | (byte2 << 6) | byte3));
+        i += 9;
+        continue;
+      }
+      if (text.charAt(i + 9) != '%') {
+        throw new IllegalArgumentException();
+      }
+      int byte4 = (digit16(text.charAt(i + 10)) << 4) + digit16(text.charAt(i + 11));
+      if ((byte4 & 0xC0) != 0x80) {
+        throw new IllegalArgumentException();
+      }
+      byte4 = byte4 & 0x3F;
+      if ((byte1 & 0xF8) == 0xF0) {
+        int codePoint = ((byte1 & 0x07) << 0x12) | (byte2 << 0x0C) | (byte3 << 0x06) | byte4;
+        if (codePoint >= 0x010000 && codePoint <= 0x10FFFF) {
+          decoded.append(Character.toChars((codePoint & 0xFFFF) >>> 10 & 0x3FF | 0xD800));
+          decoded.append(Character.toChars(0xDC00 | (codePoint & 0xFFFF) & 0x3FF));
+          i += 12;
+          continue;
+        }
+      }
+      throw new IllegalArgumentException();
+    }
+    return decoded.toString();
```

Comment:
I'd definitely appreciate it if you could scrutinize this block @theck13. I'm making some assumptions about how `URLDecoder()` works and it's worth pointing out that this entirely replaces that behavior. In the JavaScript library we're only switching over to the custom parser once the native one fails with the Unicode error.

I ran the `speedtest` and couldn't find any significant performance impact but I wasn't sure what exceptional cases could exist here.

Reply:
> I'm making some assumptions about how `URLDecoder()` works

Did you look at the [`URLDecoder.decode(String s, String enc)` source code](https://github.com/openjdk-mirror/jdk7u-jdk/blob/f4d80957e89a19a29bb9f9807d2a28351ed7f7df/src/share/classes/java/net/URLDecoder.java#L108-L206)?

> In the JavaScript library we're only switching over to the custom parser once the native one fails with the Unicode error.

Do you mean something like `URLDecoder.decode(String s, String enc)` when you refer to "the native one" above?  If so, we could do something similar by trying `URLDecoder.decode(param, "UTF-8")` first then reverting to `decodeURI(param)` if the try fails.

```java
try {
  param = URLDecoder.decode(param, "UTF-8");
} catch (UnsupportedEncodingException e) {
  param = this.decodeURI(param);
} catch (IllegalArgumentException e) {
  // Malformed URI sequence.
  throw new IllegalArgumentException(
      "Illegal escape in diff_fromDelta: " + param, e);
}
```

Reply:
I tried exploring using another decoder but didn't get anything working - it was mainly due to the combined work of percent-encoding decoding and UTF-8 decoding.

if that's working for you it'd make the patch better. we would want to `try/catch` `this.decodeURI()` as well because it can also throw upon malformed input

Reply:
> we would want to `try/catch` `this.decodeURI()` as well because it can also throw upon malformed input

I'm not sure I understand this part.  Are you saying we should add `UnsupportedEncodingException` to the second `catch` since `this.decodeURI()` could throw that exception as well?

The way that snippet above is written, the `param = this.decodeURI(param);` statement will only be run if the `param = URLDecoder.decode(param, "UTF-8");` throws an `UnsupportedEncodingException`.  Isn't that what we want?

Reply:
we have two levels of exceptions here and it's confusing.

we can percent-encode a surrogate half's _code unit_ as if it were valid Unicode and the UTF-8 algorithm can convert that value into three percent-hex-digit-pairs.

`\ud83c` for example would turn into `%ED%A0%BC` even though what we think would be the corresponding code point, `0xD83C`, is marked as entirely off-limits because it would cause confusion with surrogate pairs in UTF-16.

that being said, there's no fundamental issue in encoding that number as there is with encoding negative numbers or numbers bigger than `0x10000000`

if we had `\ud83c\udd70` as two UTF16 code units then they would represent `\U0001F70` as the code point and that would encode into four percent-hex-digit-pairs `%F0%9F%85%B0`

so this is the kind of exception we want to handle here. we specifically don't want to fail just because someone encoded half a character. we can recover from that.

on the other hand, there are structural issues we can't deal with, such as if someone sends `%123` or `%F%D%C` and there are other issues such as if they send an invalid byte sequence through the percent-hex-digit-pairs. for instance, if we got `%30` instead of `%F0` as the first pair in that encoding - `%30%9F%85%B0` - then we wouldn't be able to decode it according to the algorithm because it would specify the wrong continuation and length bits.

does that help clarify what's happening? the main point is that `this.decodeURI()` can also fail for reasons that `URLDecoder.decode()` would fail which also have nothing to do with surrogates